File: LeetCode/python/ProductOfArrayExceptSelf.py
class Solution:
    # @param {integer[]} nums
    # @return {integer[]}
    def productExceptSelf(self, nums):
        power = 1
        zeroCnt = 0
        for num in nums:
            if(num == 0):
                zeroCnt += 1
                if(zeroCnt == 2): #超过两个0的话整个数组都为0
                    power = 0
                    break
            else:		
                power *= num
								
        numsSize = len(nums)        
        if(power == 0):
            for i in range(numsSize):
                nums[i] = 0
        else:
            if(zeroCnt == 0):
                for i in range(numsSize):
                    nums[i] = power//nums[i]
            else:
                for i in range(numsSize):
                    if(nums[i] == 0):
                        nums[i] = power
                    else:
                        nums[i] = 0
			
        return nums
		
#测试程序
s = Solution()
nums = [1,0,3,5]
print(s.productExceptSelf(nums))					

File: LeetCode/iter1/c++/ProductOfArrayExceptSelf.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> productExceptSelf(vector<int>& nums)
{
    int power = 1;
	int zeroCnt = 0;
	for(int i=0;i<nums.size();i++)
	{
		if(nums[i] == 0)
		{
			zeroCnt++;
			if(zeroCnt == 2)
			{
				power=0;
				break;
			}
		}
		else
			power *= nums[i];
	}
	
	if(power == 0)
	{
		for(int i=0;i<nums.size();i++)
			nums[i] = 0;
	}
	else
	{
		if(zeroCnt == 1)
		{
			for(int i=0;i<nums.size();i++)
			{
				if(nums[i] != 0)
					nums[i] = 0;
				else
					nums[i] = power;
			}			
		}
		else
		{
			for(int i=0;i<nums.size();i++)
				nums[i] = power / nums[i];
		}
	}
	
	return nums;
}

int main()
{
	vector<int> v;
	v.push_back(1);
	v.push_back(0);
	v.push_back(3);	
	v.push_back(0);
	
	v = productExceptSelf(v);
	
	for(int i=0;i<v.size();i++)
		cout<<v[i]<<" ";
		
	return 0;
}